Admin panel navigation improvement

marcin-dykas opened the pull request:
Hello Magento Team

This is new kind of feature for admin panel. It adds additional links for products in several grids following [Customizable Options] grid.

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

Functionality for Magento 2 adds missing links in the admin panel for product edit pages. Code follows an idea from the [Customizable Options] section where is [Import options] product grid. [Import options] product grid contains URLs to product's edit page. Unfortunately, in other grids, there is no way to navigate to product edit pages easily. It adds edit URL links to all grids where navigation is needed to open product edit page in new tab.

![configurations_original](https://user-images.githubusercontent.com/4815371/129722456-3f81dee5-310d-4473-9ddb-9a9f4bfad0cc.png)
![add_related_original](https://user-images.githubusercontent.com/4815371/129722453-d5ad4e3e-c8f8-4098-be12-326048cda795.png)
![related_original](https://user-images.githubusercontent.com/4815371/129722455-311a5670-003f-463d-97cb-eee8b14b8763.png)


### Benefits

The administrator saves precious time while working with a large number of products. While adding options to products each time, there is a need to open grid with additional product items. There is no way to go to their edit pages without opening additional browser tabs and searching in other grids. It's complicated and time-consuming. So it adds edit links to the admin panel grids. After clicking in additional link product edit page is opened in a new tab.

![add_products_to_group_new](https://user-images.githubusercontent.com/4815371/129722612-4792bfce-1142-49a6-9f32-39197df26382.png)
![grouped_new](https://user-images.githubusercontent.com/4815371/129722616-243d4f0b-44d3-48d2-8995-63e863454f8d.png)
![products_to_options_new](https://user-images.githubusercontent.com/4815371/129722619-1c27788c-ca99-4608-bddc-3fa1eeb09d4d.png)
![options_new](https://user-images.githubusercontent.com/4815371/129722621-a0e42b59-3dd1-428b-b046-d5bed8a8465e.png)
![associated_new](https://user-images.githubusercontent.com/4815371/129722623-cb1cc0d4-a737-4779-b647-d6e2b8b9157a.png)
![add_related_new](https://user-images.githubusercontent.com/4815371/129722625-8f95e14f-41c5-4520-abf3-63f311330e4e.png)
![related_new](https://user-images.githubusercontent.com/4815371/129722629-8d02b04c-2a3c-45f1-a41f-7850f1f2aa1b.png)

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Open admin panel and go to Catalog -> Products
2. Edit simple product
3. Go to Related Products, Up-Sells, and Cross-Sells section
4. Click Add Related Products
5. Grid should contain links. After clicking product edit page is opened in new tab.
6. Click Up-Sell Products
7. Grid should contain links. After clicking product edit page is opened in new tab.
8. Click Cross-Sell Products
9. Grid should contain links. After clicking product edit page is opened in new tab.
10. After saving all three grids for Related, Up-Sell and Cross-Sell grids should contain links. After clicking product edit page is opened in new tab.
11. Go back to Catalog -> Products
12. Open bundle product
13. Go to Bundle Items section
14. Open Add Option -> Add Product to Option
15. Grid should contain links. After clicking product edit page is opened in new tab.
16. After saving all options grids should have links. After clicking product edit page is opened in new tab.
17. Go back to Catalog -> Products
18. Open grouped product
19. Go to Grouped Products section
20. Add Products to Group grid should contain links for products. After clicking product edit page is opened in new tab.
21. After saving Grouped Products grid should contain links for products. After clicking product edit page is opened in new tab.
22. Go back to Catalog -> Products
23. Open Configurable product
24. Go to Configurations section
25. Add Products Manually product grid should contain links. After clicking product edit page is opened in new tab.
26. After saving Configurations grid should contain links. After clicking product edit page is opened in new tab.
27. Go to Customizable Options section
28. Import Options grid should contain links. After clicking product edit page is opened in new tab.



### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

mdykas-gorilla committed:
Admin panel navigation improvement

bgorski commented:
@marcin-dykas thank you for your contribution! However, before can proceed with reviewing the code and prioritizing the PR, you need to sign the Adobe Contributor License agreement. Can you do that please? You can find it here: https://opensource.adobe.com/cla.html

marcin-dykas commented:
Hello @bgorski 

I've just signed CLA.

Thanks.

bgorski commented:
@magento run all tests

BarnyShergold commented on AddUrlToName.php:
```diff
@@ -0,0 +1,105 @@
+<?php
+
```
Missing copyright notice

marcin-dykas replied:
Fixed by adding copyright notice

BarnyShergold commented on ConfigurableDataProvider.php:
```diff
@@ -0,0 +1,50 @@
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\ConfigurableProduct\Ui\DataProvider\Product;
```
Missing - declare(strict_types=1);

marcin-dykas replied:
Fixed by adding declare(strict_types=1);

BarnyShergold commented on ConfigurableDataProvider.php:
```diff
@@ -0,0 +1,50 @@
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory;
+use Magento\Catalog\Ui\DataProvider\Product\ProductDataProvider;
+use Magento\Ui\DataProvider\Modifier\PoolInterface;
+
```
Missing class DOCBLOCK

marcin-dykas replied:
Fixed by adding DOCBLOCK

BarnyShergold commented on app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/Composite.php:
```diff
@@ -45,24 +46,32 @@ class Composite extends AbstractModifier
+    /**
+     * @var NameHelper
+     */
+    protected $nameHelper;
```
should be private

marcin-dykas replied:
Fixed by changing to private

BarnyShergold commented on app/code/Magento/Catalog/Helper/Product/AddUrlToName.php:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * @var LocatorInterface
+     */
+    protected $locator;
```
class variables should be private

marcin-dykas replied:
Fixed by changing to private

BarnyShergold commented on app/code/Magento/Catalog/Helper/Product/AddUrlToName.php:
```diff
@@ -0,0 +1,72 @@
+
+        $url = $this->urlBuilder->getUrl(
+            'catalog/product/edit',
+            ['id' => $linkedProduct->getId(), 'store' => $storeId]
```
Ideally one element per line

marcin-dykas replied:
Fixed by moving to separate lines

BarnyShergold commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Related.php:
```diff
@@ -98,6 +99,11 @@ class Related extends AbstractModifier
+    /**
+     * @var NameHelper
+     */
+    protected $nameHelper;
```
private

marcin-dykas replied:
Fixed by changing to private

BarnyShergold commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Related.php:
```diff
@@ -706,4 +715,36 @@ protected function getTextColumn($dataScope, $fit, Phrase $label, $sortOrder)
+     * @param int $sortOrder
+     * @return array
+     */
+    protected function getHtmlColumn($dataScope, $fit, Phrase $label, $sortOrder)
```
missing return type on prototype
should be private

marcin-dykas replied:
Fixed by changing to private. Added return type

BarnyShergold commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Modifier/AddUrlToName.php:
```diff
@@ -0,0 +1,105 @@
+<?php
+
```
missing copyright

marcin-dykas replied:
Fixed by adding copyright

BarnyShergold commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Related/AbstractDataProvider.php:
```diff
@@ -205,4 +214,28 @@ protected function getStore()
+     *
+     * @return array
+     */
+    public function getData()
```
missing return type on prototype

marcin-dykas replied:
Fixed by adding return type

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Ui/DataProvider/Product/ConfigurableDataProvider.php:
```diff
@@ -0,0 +1,50 @@
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\ConfigurableProduct\Ui\DataProvider\Product;
```
missing declare(strict_types=1);

marcin-dykas replied:
Fixed by adding declare(strict_types=1);

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Ui/DataProvider/Product/ConfigurableDataProvider.php:
```diff
@@ -0,0 +1,50 @@
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory;
+use Magento\Catalog\Ui\DataProvider\Product\ProductDataProvider;
+use Magento\Ui\DataProvider\Modifier\PoolInterface;
+
```
Missing class DOCBLOCK

marcin-dykas replied:
Fixed by adding DOCBLOCK

BarnyShergold commented on app/code/Magento/GroupedProduct/Ui/DataProvider/Product/Form/Modifier/Grouped.php:
```diff
@@ -112,6 +113,11 @@ class Grouped extends AbstractModifier
+    /**
+     * @var NameHelper
+     */
+    protected $nameHelper;
```
private

marcin-dykas replied:
Fixed by changing to private

BarnyShergold commented on app/code/Magento/GroupedProduct/Ui/DataProvider/Product/Form/Modifier/Grouped.php:
```diff
@@ -670,4 +679,36 @@ protected function getTextColumn($dataScope, $fit, Phrase $label, $sortOrder)
+     * @param int $sortOrder
+     * @return array
+     */
+    protected function getHtmlColumn($dataScope, $fit, Phrase $label, $sortOrder)
```
missing return type

marcin-dykas replied:
Fixed by adding return type

BarnyShergold reviewed (changes requested):
Good submission!

However there are a lot of failed tests in the areas this PR covers. Please check and fix the failed tests. 

There are quite a few new classes and new methods on existing classes yet there are no Unit Tests or MFTF tests. Please look into how your new functionality can be covered by testing.

mdykas-gorilla committed:
Removed wrong files

mdykas-gorilla committed:
Fixes after CR

marcin-dykas committed:
Merge branch 'magento:2.4-develop' into feature/improve-admin-panel-navigation

mdykas-gorilla committed:
Adjusted existing unit tests

mdykas-gorilla committed:
Adjusted existing unit tests - added missing file

mdykas-gorilla committed:
Adjusted existing unit tests - added missing file

mdykas-gorilla committed:
Adjusted existing unit tests - added test file for ConfigurableDataProvider

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/ConfigurableDataProviderTest.php:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * @var ObjectManager
+     */
+    protected $objectManager;
```
All protected should be private

marcin-dykas replied:
Fixed by changing all protected to private

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/ConfigurableDataProviderTest.php:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * @return object
+     */
+    protected function getModel()
```
missing return type
should be private

marcin-dykas replied:
Fixed by adding return type and changed to private

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/ConfigurableDataProviderTest.php:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Test checks ConfigurableDataProvider type
+     */
+    public function testCheckType()
```
missing return type

marcin-dykas replied:
Fixed by adding return type

BarnyShergold commented on app/code/Magento/ConfigurableProduct/Test/Unit/Ui/DataProvider/Product/ConfigurableDataProviderTest.php:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Test checks collection type
+     */
+    public function testGetCollection()
```
missing return type

BarnyShergold replied:
Usually it's good practice to put all the public methods first  and then the private ones

marcin-dykas replied:
Fixed by moving public functions first. Then protected then private ones at the end ....

BarnyShergold reviewed (changes requested):
Great test coverage from what I can see.
Just a few tidy ups

BarnyShergold commented:
@marcin-dykas - It\'s been 2 weeks since changes were requested on this PR for the tests and there is also a conflict to resolve. Can you please give us an update on progress and when you are likely to be able to complete this PR? Otherwise we may have to look at closing it. And we don't want to do that!

mdykas-gorilla committed:
Added new unit tests for AddUrlToName method

marcin-dykas commented:
Hello @BarnyShergold 

I've just added last missing unit tests. Now I'll tidy up things to run tests again here.

mdykas-gorilla committed:
Fixes after last CR

mdykas-gorilla committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into feature/improve-admin-panel-navigation

marcin-dykas commented:
@magento run Unit Tests

marcin-dykas commented:
I've run unit tests here and there are two same errors from Jsunit (B2B) and Jsunit (EE)

Interception cache generation... 6/9 [========>----]  66% 54 secs 400.0 MiBErrors during compilation:
	Magento\PricePermissions\Ui\DataProvider\Product\Form\Modifier\Grouped
		Missed required argument nameHelper in parent::__construct call. File: /var/www/html/app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Grouped.php
	Magento\PricePermissions\Ui\DataProvider\Product\Form\Modifier\Related
		Incompatible argument type: Required type: \Magento\Catalog\Helper\Product\AddUrlToName. Actual type: string; File:
/var/www/html/app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Related.php

This is because EE version has Magento_PricePermissions which is not present in CE. 

In this module there are classes like app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Grouped.php and app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Related.php. 

They extend \Magento\GroupedProduct\Ui\DataProvider\Product\Form\Modifier\Grouped and \Magento\Catalog\Ui\DataProvider\Product\Form\Modifier\Related where I added one new parameter to constructor.

It should be easy to fix these errors. We just have to add new parameters in parent::_construct calls.

I don't have access to EE repository and I wonder now how to solve it. @BarnyShergold what is procedure in such case ? Is there any ? We'll leave it as is or we are going to create PR to EE repo ? Could you please advise ?

Thank you

mdykas-gorilla committed:
More fixes after last CR

mdykas-gorilla committed:
Merge remote-tracking branch 'upstream/2.4-develop' into feature/improve-admin-panel-navigation

mdykas-gorilla committed:
Fixed problem with undeclared dependencies

mdykas-gorilla committed:
Fixed error from Magento\Test\Php\LiveCodeTest::testCodeStyle

mdykas-gorilla committed:
Fixed error from Magento\Test\Php\LiveCodeTest::testCodeStyle

mdykas-gorilla committed:
Fixed error in Magento\ConfigurableProduct\Ui\DataProvider\Product\Form\Modifier\Data\AssociatedProductsTest::testAddManuallyConfigurationsWithNotFilterableInGridAttribute

mdykas-gorilla committed:
Fixed error in Magento\Test\Php\LiveCodeTest::testPhpStan

app/code/Magento/Catalog/Ui/DataProvider/Product/Related/AbstractDataProvider.php

app/code/Magento/GroupedProduct/Ui/DataProvider/Product/GroupedProductDataProvider.php

marcin-dykas commented:
@magento run all tests

marcin-dykas commented:
Still 9 checks are not successful but most of them fails because EE PricePermission module.

Error says:
```
Interception cache generation... 6/9 [========>----]  66% 1 min 444.0 MiBErrors during compilation:
	Magento\PricePermissions\Ui\DataProvider\Product\Form\Modifier\Grouped
		Missed required argument nameHelper in parent::__construct call. File: /var/www/html/app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Grouped.php
	Magento\PricePermissions\Ui\DataProvider\Product\Form\Modifier\Related
		Incompatible argument type: Required type: \Magento\Catalog\Helper\Product\AddUrlToName. Actual type: string; File:
/var/www/html/app/code/Magento/PricePermissions/Ui/DataProvider/Product/Form/Modifier/Related.php
```

Since I don't have access to EE repository I can't fix it. I decided to provided patch to this module which solves this issue.


[EE_PricePermissions_module_patch.zip](https://github.com/magento/magento2/files/7625969/EE_PricePermissions_module_patch.zip)

mdykas-gorilla committed:
Merge remote-tracking branch 'upstream/2.4-develop' into feature/improve-admin-panel-navigation
